lib/ChallengeManager.php

```diff
@@ -2,34 +2,166 @@
             Logger::error(self::LOG_PREFIX . 'Error while creating a challenge');
             http_response_code(500);
-            return false;
+            throw new Exception('ChallengeManager.generateChallenge: Error while generating a challenge');
+        }
+        return $challenge;
+    }
+
+    public function generateToken($id, $scriptName, $data)
+    {
+
+        $challenge = $this->generateChallenge($id, $scriptName);
+
+        if (empty($challenge)) {
+            Logger::error('Retrieving the challenge was not successful.');
+            return;
         }
 
-        return true;
+        $jwk = JWKFactory::createFromKeyFile($this->privKey);
+        $algorithmManager = new AlgorithmManager(
+            [
+                self::getAlgorithm('Signature\\Algorithm', $this->sighAlg)
+            ]
+        );
+        $jwsBuilder = new JWSBuilder($algorithmManager);
+        $payload = json_encode([
+            'iat' => time(),
+            'nbf' => time(),
+            'exp' => time() + 300,
+            'challenge' => $challenge,
+            'id' => $id,
+            'data' => $data
+        ]);
+
+        $jws = $jwsBuilder
+            ->create()
+            ->withPayload($payload)
+            ->addSignature($jwk, ['alg' => $this->sighAlg])
+            ->build();
+
+        $serializer = new CompactSerializer();
+        $token = $serializer->serialize($jws, 0);
+
+        return $token;
     }
 
-    public function readChallengeFromDb($id)
+    public function decodeToken($token)
+    {
+        $algorithmManager = new AlgorithmManager(
+            [
+                self::getAlgorithm('Signature\\Algorithm', $this->sighAlg)
+            ]
+        );
+        $jwsVerifier = new JWSVerifier($algorithmManager);
+        $jwk = JWKFactory::createFromKeyFile($this->pubKey);
+
+        $serializerManager = new JWSSerializerManager([new CompactSerializer()]);
+        $jws = $serializerManager->unserialize($token);
+
+        $headerCheckerManager = new HeaderCheckerManager(
+            [new AlgorithmChecker([$this->sighAlg])],
+            [new JWSTokenSupport()]
+        );
+
+        $headerCheckerManager->check($jws, 0);
+
+        $isVerified = $jwsVerifier->verifyWithKey($jws, $jwk, 0);
+
+        if (!$isVerified) {
+            Logger::error('Perun.updateUes: The token signature is invalid!');
+            http_response_code(401);
+            exit;
+        }
+
+        $claimCheckerManager = new ClaimCheckerManager(
+            [
+                new Checker\IssuedAtChecker(),
+                new Checker\NotBeforeChecker(),
+                new Checker\ExpirationTimeChecker(),
+            ]
+        );
+        $claims = json_decode($jws->getPayload(), true);
+
+        $claimCheckerManager->check($claims);
+
+        $challenge = $claims['challenge'];
+        $id = $claims['id'];
+
+        $challengeManager = new ChallengeManager();
+
+        $challengeDb = $challengeManager->readChallengeFromDb($id);
```

Comment:
Challenge is saved with scriptName but read without scriptName. Either remove the scriptName or use it for reading as well.